Modules/_datetimemodule.c

```diff
@@ -6921,70 +7076,148 @@ init_state(datetime_state *st, PyTypeObject *PyDateTime_IsoCalendarDateType)
+    Py_VISIT(st->isocalendar_date_type);
 
     return 0;
 }
 
 static int
-_datetime_exec(PyObject *module)
+clear_state(datetime_state *st)
 {
+    Py_CLEAR(st->isocalendar_date_type);
+    Py_CLEAR(st->us_per_ms);
+    Py_CLEAR(st->us_per_second);
+    Py_CLEAR(st->us_per_minute);
+    Py_CLEAR(st->us_per_hour);
+    Py_CLEAR(st->us_per_day);
+    Py_CLEAR(st->us_per_week);
+    Py_CLEAR(st->seconds_per_day);
+    Py_CLEAR(st->epoch);
+    return 0;
+}
+
+
+/* ---------------------------------------------------------------------------
+ * Global module state.
+ */
+
+// If we make _PyStaticType_*ForExtension() public
+// then all this should be managed by the runtime.
+
+static struct {
+    PyMutex mutex;
+    int64_t interp_count;
+} _globals = {0};
+
+static void
+callback_for_interp_exit(void *Py_UNUSED(data))
+{
+    PyInterpreterState *interp = PyInterpreterState_Get();
+
+    assert(_globals.interp_count > 0);
+    PyMutex_Lock(&_globals.mutex);
+    int64_t final = !_globals.interp_count;
+    _globals.interp_count -= 1;
+    PyMutex_Unlock(&_globals.mutex);
+
+    for (size_t i = 0; i < Py_ARRAY_LENGTH(capi_types); i++) {
+        PyTypeObject *type = capi_types[i];
+        _PyStaticType_FiniForExtension(interp, type, final);
+    }
+}
+
+static int
+init_static_types(PyInterpreterState *interp, int reloading)
+{
+    if (reloading) {
+        return 0;
+    }
+
     // `&...` is not a constant expression according to a strict reading
     // of C standards. Fill tp_base at run-time rather than statically.
     // See https://bugs.python.org/issue40777
     PyDateTime_TimeZoneType.tp_base = &PyDateTime_TZInfoType;
     PyDateTime_DateTimeType.tp_base = &PyDateTime_DateType;
 
-    PyTypeObject *capi_types[] = {
-        &PyDateTime_DateType,
-        &PyDateTime_DateTimeType,
-        &PyDateTime_TimeType,
-        &PyDateTime_DeltaType,
-        &PyDateTime_TZInfoType,
-        &PyDateTime_TimeZoneType,
-    };
-
     for (size_t i = 0; i < Py_ARRAY_LENGTH(capi_types); i++) {
-        if (PyModule_AddType(module, capi_types[i]) < 0) {
-            goto error;
+        PyTypeObject *type = capi_types[i];
+        if (_PyStaticType_InitForExtension(interp, type) < 0) {
+            return -1;
         }
     }
 
-#define CREATE_TYPE(VAR, SPEC, BASE)                    \
-    do {                                                \
-        VAR = (PyTypeObject *)PyType_FromModuleAndSpec( \
-                module, SPEC, (PyObject *)BASE);        \
-        if (VAR == NULL) {                              \
-            goto error;                                 \
-        }                                               \
-    } while (0)
+    PyMutex_Lock(&_globals.mutex);
+    assert(_globals.interp_count >= 0);
+    _globals.interp_count += 1;
+    PyMutex_Unlock(&_globals.mutex);
 
-    PyTypeObject *PyDateTime_IsoCalendarDateType = NULL;
-    datetime_state *st = get_datetime_state();
+    // XXX Add one callback per type?
```

Comment:
Nit: change `XXX` to `FIXME` or `TODO` and make a note of this in the issue, for visibility.

Reply:
Ultimately this comment was more of an implementation note and a short-term TODO.  I'll take care of it.  Thanks for pointing it out.

Reply:
done